keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+above this is the same `Capacity` type, but more complex models will have
+a more complex capacity type and a simpler parameter type.
+
+When defining the CRD for the claim parameters, the driver developer replicates
+those fields from the numeric model which are relevant for the driver and adds
+custom validation. For a complex model, this can be a true subset of the fields
+in the parameter type. The validation can set bounds that are specific to the
+driver. One can think of this as "duck typing": the CRD doesn't need to include
+the numeric type, a subset of it just needs to behave like one.
+
+To enable extracting the numeric parameters, they must be located at a specific
+JSON path (e.g. "spec"). "Kind" and "apiVersion" must be included at that
+path. Users don't need to set those, the CRD can add them via defaults. This is
+best explained with an example:
+
+```
+const GroupName = "dra.e2e.example.com"
+const KindName = "ClaimParameter"
+const Version = "v1alpha1"
+
+type ClaimParameter struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta   `json:",inline"`
+
+	Spec ClaimParameterSpec `json:"spec"`
+}
+
+type ClaimParameterSpec struct {
+	TypeMeta `json:",inline"` // Not metav1.TypeMeta! See below.
+
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+    // +kubebuilder:validation:Maximum=16
+	Count int64 `json:"count,omitempty"`
+
+	Env map[string]string `json:"env,omitempty"`
+}
+
+type TypeMeta struct {
+	// +kubebuilder:default=Capacity
+	// +kubebuilder:validation:Enum=Capacity
+	Kind string `json:"kind,omitempty"`
+
+	// +kubebuilder:default=counter.dra.config.k8s.io/v1alpha1
+	// +kubebuilder:validation:Enum=counter.dra.config.k8s.io/v1alpha1
+	APIVersion string `json:"apiVersion,omitempty"`
+}
+```
+
+In this example, the driver supports the `Count` numeric parameter and adds
+environment variables as configuration parameters. There is a 1:1 relationship
+between the CRD type and the numeric type that it is using. Users define
+parameters like they would without numeric parameters:
+
+```
+kind: ClaimParameter
+apiVersion: dra.e2e.example.com
+spec:
+  count: 10
+  env:
+    HELLO_WORLD: "Thanks for the fish."
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list and
+watch arbitrary resources. This is necessary because it is not known in advance
+which CRDs will be used by DRA drivers.
+
+This may change at runtime as classes get added or removed. To avoid repeated
+GET operations for those CRDs, dynamic informers will have to be used that get
+started and stopped as needed.
+
+### Risks and Mitigations
+
+Granting kube-scheduler permission to read arbitrary types could be used by an
+attacker to extract information. However, for that an attacker first would have
+to gain control of the scheduler, get it to execute API calls, and then
+retrieve the result. This seems like a low risk.
```

Comment:
This is moot given the discussion above about DRA drivers bringing along their own role / role-binding, correct?

Reply:
Correct.